### ISSUE TITLE
Hovering Wing Deficiencies

### ISSUE THREAD

**Issue by Antirhinnum** [bluemagic123/tModLoader#3441]
### Do you intend to personally contribute/program this feature?

Yes

### I would like to see this change made to improve my experience with

Mod capability as a Modder

### Description

Some way to mark that a modded pair of wings is able to hover, similar to vanilla's Hoverboard. Additionally, some way to specify that a pair of wings should consume less flight time when hovering and not moving. Both of these are hardcoded for specific vanilla wings.

In short, modders would (ideally) be able to specify the following for modded wings:
* Whether or not the wings should allow hovering.
* The slowdown factor of entering the hovering state (see "Other solutions").
* The flight time usage per frame while hovering in place.

These would be implemented using sets in `ArmorIDs.Wing.Sets`:
* `ArmorIDs.Wing.Sets.CanHover` (default: false)
* `HoverSlowdown` (default: 0.9f)
* `HoverFlightTimeCost` (default: 0.5f)

### What does this proposal attempt to solve or improve?

Currently, there is no reasonable way to make a modded pair of wings hover, despite the functionality being seen in many vanilla wings (Hoverboard, Vortex Booster, Celestial Starboard, etc.).

`ArmorIDs.Wing.Sets.Stats` and the associated `WingStats` struct allow specifying how fast a player should move when trying to hover. This behavior works fine, but does not actually make the player hover: They just speed up if holding down (or equivalent).

The code that handles hovering is right in the middle of `Player.Update()`, so it's unreasonable to IL edit in hovering logic, especially for something so common in vanilla wings. Additionally, there are no nearby tModLoader hooks where a modder could handle hovering logic manually.

It *is* currently reasonable to make a pair of wings consume less flight time when "hovering" using `Mod/GlobalItem.VerticalWingSpeeds()` by directly modifying `Player.wingTime` to offset the default `-1f` per frame. However, this approach is unintuitive and must be done defensively to avoid the modder increasing the player's wing time infinitely. It also doesn't work for the Jetpack's "boost" ability, but that's its own can of worms.

### Which (other) solutions should be considered?

The blocks of code in vanilla that handle hovering and flight time use are as follows:
```cs
// void Player.Update()
if ((wingsLogic == 22 || wingsLogic == 28 || wingsLogic == 30 || wingsLogic == 31 || wingsLogic == 33 || wingsLogic == 35 || wingsLogic == 37 || wingsLogic == 45) && TryingToHoverDown && controlJump && wingTime > 0f && !merman) {
	float num71 = 0.9f;
	if (wingsLogic == 45)
		num71 = 0.8f;

	velocity.Y *= num71;
	if (velocity.Y > -2f && velocity.Y < 1f)
		velocity.Y = 1E-05f;
}

if (wingsLogic == 37 && TryingToHoverDown && controlJump && wingTime > 0f && !merman) {
	velocity.Y *= 0.92f;
	if (velocity.Y > -2f && velocity.Y < 1f)
		velocity.Y = 1E-05f;
}
```
```cs
// Player.WingMovement()
if ((wingsLogic == 22 || wingsLogic == 28 || wingsLogic == 30 || wingsLogic == 31 || wingsLogic == 37 || wingsLogic == 45) && TryingToHoverDown && !controlLeft && !controlRight)
	wingTime -= 0.5f;
else
	wingTime -= 1f;
```
The "slowdown factor" mentioned in the main proposal are the `0.9f`, `0.8f`, and `0.92f` values in `Player.Update()` that determine how quickly the player slows down when entering a hovering state.

As is always the case with new sets, these will require maintenance if any future *Terraria* updates add more wings, which not be desired. However, only keeping track of these stats for modded wings would make it incredibly difficult for modders to programmatically check these fields on vanilla wings, an issue that already exists with vanilla grappling hooks.

Raw sets for the proposed fields could work, but may be cumbersome compared to `WingStats`. Turning these into fields or properties on `Item` or `ModItem` seems excessive, as these fields have no relevance to non-wing items.

The `WingStats` struct is a candidate for implementation, but not preferred. The struct already has a `HasHoverDownStats` property, which decides if a pair of wings changes speed while hovering. Amending this would require either reassigning the value to all vanilla wings (not easily maintainable) or making a second property (might be confusing for modders due to similar name). The `HasHoverDownStats` bool could also be renamed to something less confusing, likely through hiding and a property (like how tModLoader handles `Item.melee`).

**Comment by Chicken-Bones** [comment]
I'm always a fan of more `Sets`. Configuration is king! Please refactor the vanilla code to use the new sets you add.

**Comment by AdonXCX** [comment]
Yeah I tried to replicate the behavior of the Hoverboard but use a custom graphic and I cannot for the life of me get my custom hoverboard graphic to load in, having a way to make a "hoverboard class" wing item would be useful. Using hoverboard item defaults guarantees it loads in the default hoverboard graphic.
